modules/pam_lastlog/pam_lastlog.c

```diff
@@ -278,10 +280,16 @@ last_login_read(pam_handle_t *pamh, int announce, int last_fd, uid_t uid, time_t
     last_lock.l_start = sizeof(last_login) * (off_t) uid;
     last_lock.l_len = sizeof(last_login);
 
-    if (fcntl(last_fd, F_SETLK, &last_lock) < 0) {
+    while (fcntl(last_fd, F_SETLK, &last_lock) < 0) {
+        if (0 == --lock_retries) {
+            D(("locking %s failed", _PATH_LASTLOG));
+            pam_syslog(pamh, LOG_ERR,
+                       "file %s is locked/read", _PATH_LASTLOG);
+            return PAM_SERVICE_ERR;
+        }
         D(("locking %s failed..(waiting a little)", _PATH_LASTLOG));
-	pam_syslog(pamh, LOG_WARNING,
-		   "file %s is locked/read", _PATH_LASTLOG);
+        pam_syslog(pamh, LOG_WARNING,
+                   "file %s is locked/read, retrying", _PATH_LASTLOG);
```

Comment:
I'd say that in the new context the message about retrying is not a warning anymore,
so LOG_INFO would be more suitable.